[Beta] docs(cn): translate reference/render.md (#632)
reactjs/zh-hans.reactjs.org: beta/src/pages/reference/render.md

---
title: render()
translators:

>   - liu-jin-yi

Reviewer: ```suggestion
  - liu-jin-yi
  - QC-L
```

---

<Intro>

> “render”可以将一段 [JSX](/learn/writing-markup-with-jsx)（“React 元素”）渲染到浏览器 DOM 容器节点中。它可以让 React 改变“container”内的 DOM ，使其与传递的 JSX 相匹配。

Reviewer: ```suggestion
`render` 函数会将 [JSX](/learn/writing-markup-with-jsx)（“React 元素”）渲染到浏览器 DOM 容器节点中。它可以让 React 改变 `container` 中 DOM，使其与传递的 JSX 相匹配。
```

Reviewer: 不要改变原文的符号

</Intro>

## 渲染根组件 {/*rendering-the-root-component*/}

> 想要调用“render”，你需要一段 JSX 和一个 DOM 容器：

Reviewer: ```suggestion
如需调用 `render`，需要编写一段 JSX 代码以及一个 DOM 容器：
```

<APIAnatomy>

<AnatomyStep title="React element">

> 你想渲染的 UI 界面。

Reviewer: ```suggestion
需要渲染的 UI 界面。
```

</AnatomyStep>

<AnatomyStep title="DOM container">

> 你如果想把你的 UI 界面渲染到 DOM 节点中去。那么容器是不能被修改的，只能修改它的子节点。

Reviewer: ```suggestion
用于渲染 UI 界面的 DOM 节点。该容器不能被修改，只能修改它的子节点。
```

</APIAnatomy>

> 在完全由 React 构建的应用程序中，你将在你的应用程序的顶层做一次这样的工作 —— 渲染"root"组件。

Reviewer: ```suggestion
在完全由 React 构建的应用程序中，你需要在你的应用程序的入口文件执行该操作（渲染 "root" 组件）。
```

```js App.js
export default function App() {
  return <h1>你好，世界！</h1>;
}
```

</Sandpack>

<br />

## 渲染多个根组件 {/*rendering-multiple-roots*/}

> 不管你在哪里使用 React 的 ["sprinkles"](/learn/add-react-to-a-website) ，你都应该为每个由 React 管理的顶层 UI 组件调用“render”进行渲染。

Reviewer: ```suggestion
不管你用  ["哪种方式"](/learn/add-react-to-a-website)  使用 React，你都需为每个由 React 管理的顶层 UI 组件调用 `render` 函数进行渲染。
```

<Sandpack>

```html public/index.html
<nav id="navigation"></nav>
<main>

>   <p>这一段没有被 React 渲染（可以打开 index.html 来验证）。</p>

Reviewer: ```suggestion
  <p>这一段没有被 React 渲染（可以查看 index.html 进行验证）。</p>
```

<br />

## 更新已渲染的 DOM {/*updating-the-rendered-tree*/}

> 你可以在同一个 DOM 节点上多次调用“render”。只要组件树结构与之前渲染的内容一致，React 就会 [保留这个状态](/learn/preserving-and-resetting-state) 。请注意你书写代码的方式：

Reviewer: ```suggestion
你可以在同一 DOM 节点上多次调用 `render`。只要组件树结构与之前渲染的内容一致，React 就会 [保留该状态](/learn/preserving-and-resetting-state) 。请仔细观察在输入框中输入内容后的变化：
```

Reviewer: 这里你理解错了

<br />

> ## 更新已渲染的 DOM {/*updating-the-rendered-tree*/}

Reviewer 2: 建议改成 更新已渲染的树 或 更新已渲染的组件树，因为组件树是 React 中的，DOM 树是浏览器中的，两者不能混淆，而且下文中有对两者关系的叙述，应该区别清楚，避免增加理解难度

  return (
    <>
      <h1>Hello, world! {counter}</h1>
      <input placeholder="在这里输入一些东西" />
    </>
  );
}
```

</Sandpack>

> 你可以使用 [“unmountComponentAtNode()”](TODO) 销毁已渲染的 DOM 树。

Reviewer: ```suggestion
你可以使用 [`unmountComponentAtNode()`](TODO) 来销毁已渲染的 DOM 树。
```

<br />

## 何时不使用它 {/*when-not-to-use-it*/}

> * 如果你的应用程序使用服务器渲染并在服务器上生成 HTML ，请使用 [“hydrate”](TODO) 而不是“render”。

Reviewer: ```suggestion
* 如果你的应用程序使用服务器渲染，并会在服务器上生成 HTML，请使用 [`hydrate`](TODO) 函数，而非 `render` 函数。
```

> * 如果你的应用程序是完全用 React 构建的，你应该不需要多次使用“render”。如果你想在 DOM 树的不同部分渲染一些东西（例如，modal 或者 tooltip），那么请使用 [“createPortal”](TODO) 来代替。

Reviewer: ```suggestion
* 如果你的应用程序完全基于 React 构建，你大概率不需要多次使用 `render` 函数。如果你想在 DOM 树的其他位置渲染内容（例如，modal 或者 tooltip），那么请使用 [`createPortal`](TODO) 来代替。
```

<br />

> ## 使用细节 {/*behavior-in-detail*/}

Reviewer: ```suggestion
## 行为细节 {/*behavior-in-detail*/}
```

> 在你第一次调用“render”时，“container”内的任何现有 DOM 元素都会被替换。如果你再次调用“render”，React 将会通过 ["匹配"](/learn/preserving-and-resetting-state) 与先前渲染的组件树来决定 DOM 的哪些部分可以重用，哪些需要重新创建。重复调用“render”与调用“setState” —— 在这两种情况下，React 都会避免不必要的 DOM 更新。

Reviewer: ```suggestion
在你第一次调用 `render` 时，`container` 内的任何已有 DOM 元素都会被替换。如果你再次调用 `render` 时，React 将会通过与先前渲染的组件树 ["匹配"](/learn/preserving-and-resetting-state) 的方式，来决定 DOM 的哪些部分可以重用，哪些需要重新创建。重复调用 `render` 与调用 `setState` 效果类似。无论哪种情况，React 都会避免不必要的 DOM 更新。
```

> 你可以传递一个回调函数作为第三个参数。React 会在你的组件出现在 DOM 中后调用它。

Reviewer: ```suggestion
你可以将一个回调函数，作为 `render` 的第三个参数。React 会在你的组件在 DOM 中出现后，调用它。
```

> 如果你渲染 `<MyComponent />` ，并且“MyComponent”是一个类组件，“render”将返回该类的实例。在其他情况下，它将返回“null”。

Reviewer: ```suggestion
如果需要渲染 `<MyComponent />` ，并且 `MyComponent` 是一个类组件，`render` 函数将返回该类的实例。在其他情况下，它将返回 `null`。
```

  return (
    <>
      <h1>Hello, world! {counter}</h1>
      <input placeholder="在这里输入一些东西" />
    </>
  );
}
```

</Sandpack>

你可以使用 [`unmountComponentAtNode()`](TODO) 来销毁已渲染的 DOM 树。

<br />

## 何时不使用它 {/*when-not-to-use-it*/}

* 如果你的应用程序使用服务器渲染，并会在服务器上生成 HTML，请使用 [`hydrate`](TODO) 函数，而非 `render` 函数。
* 如果你的应用程序完全基于 React 构建，你大概率不需要多次使用 `render` 函数。如果你想在 DOM 树的其他位置渲染内容（例如，modal 或者 tooltip），那么请使用 [`createPortal`](TODO) 来代替。

<br />

## 行为细节 {/*behavior-in-detail*/}

> 在你第一次调用 `render` 时，`container` 内的任何已有 DOM 元素都会被替换。如果你再次调用 `render` 时，React 将会通过与先前渲染的组件树 ["匹配"](/learn/preserving-and-resetting-state) 的方式，来决定 DOM 的哪些部分可以重用，哪些需要重新创建。重复调用 `render` 与调用 `setState` 效果类似。无论哪种情况，React 都会避免不必要的 DOM 更新。

Reviewer 2: 如果你再次调用 `render` 的话，React 会为了体现最新的 JSX 而进行必要的 DOM 更新。React 会通过将 DOM 与先前渲染的组件树进行 ["匹配"](/learn/preserving-and-resetting-state) 的方式，来决定 DOM 的哪些部分可以重用，哪些需要重新创建。重复调用 `render` 与调用 `setState` 类似——在这两种情况下，React 都会避免不必要的 DOM 更新。

1. 这里”React will update the DOM as necessary to reflect the latest JSX.“漏译了
2. matching it up 中的 it 指的是 DOM，应该翻译出来
3. 建议把”效果类似“中的”效果“去掉。不应该说两种调用的效果类似，因为本质上 render 是负责挂载组件的而 setState 是负责改变 state 的。只能说在避免不必要的 DOM 更新方面两者是类似的。

> 你可以将一个回调函数，作为 `render` 的第三个参数。React 会在你的组件在 DOM 中出现后，调用它。

Reviewer 2: 你可以将一个回调函数作为第三个参数传递给 `render`。React 会在你的组件在 DOM 中出现后调用它。

1. pass 漏译了
2. 谓语之前不要加逗号
